plane/src/drone/executor.rs

```diff
@@ -53,13 +61,77 @@ impl<R: Runtime> Executor<R> {
 
         Self {
             runtime,
-            state_store: Arc::new(Mutex::new(state_store)),
+            state_store,
             backends,
             ip,
             _backend_event_listener: backend_event_listener,
         }
     }
 
+    // On restart, we want to terminate all existing backends and start fresh.
+    // This prevents bugs where an agent restart leaves the drone unable to
+    // terminate old backends.
+    async fn terminate_preexisting_backends(
+        runtime: Arc<R>,
+        state_store: Arc<Mutex<StateStore>>,
+    ) -> Result<()> {
+        let backends = state_store
+            .lock()
+            .expect("State store lock poisoned.")
+            .active_backends()?;
+
+        tracing::info!(?backends, "Terminating preexisting backends");
+
+        let mut tasks = vec![];
+        let state_store_clone = state_store.clone();
+        for (backend_id, state) in backends {
+            let runtime_clone = runtime.clone();
+            let state_store = state_store_clone.clone();
+            let state = state.clone();
+            tasks.push(tokio::spawn(async move {
+                let mut backoff = ExponentialBackoff::default();
+                let mut success = false;
+                for attempt in 1..=10 {
+                    match runtime_clone.terminate(&backend_id, true).await {
+                        Ok(()) => {
+                            success = true;
+                            break;
+                        }
+                        Err(err) => {
+                            tracing::error!(
```

Comment:
If the container is not found, I'm assuming that results in an `Err`? We should probably also count that as a success and update the backend's state to `Terminated`.